### PR TITLE
BFD-1846: Create New Cloudwatch metrics

### PR DESCRIPTION
**Pull request by kcirtapfromspace**
**JIRA Ticket:**
[BFD-1846](https://jira.cms.gov/browse/BFD-1846)

**User Story or Bug Summary:**

User Story: As a BFD Engineer I do not want the Cloudwatch dashboard to use a deprecated log source.

---

### What Does This PR Do?

This PR addresses defining metrics filters for the access.json logs. Existing filters for access.txt are moved to a new terraform resource where objects are cataloged with a notation of txt within the metric name and resources. Access.json filters are applied to the same namespace using previous metric nomenclature. 

This should address existing databoards using depricated log sources while providing verification that both metric types do not differ from each other. 

Validations: 
 - Checked that all metrics land in the same metric namespace 
  ```
  aws cloudwatch --profile bfd list-metrics --namespace bfd-test/bfd-server | jq '.Metrics[].MetricName'
"http-requests/count-500/bb"
"http-requests/count/eobAll/ab2d"
"http-requests/count/coverageAll/all"
"http-requests/count/coverageAll/ab2d"
"http-requests/count-500/all"
"http-requests/count-not-2xx/ab2d"
"http-requests/latency/all/all"
"http-requests/count/patientAll/bcda"
"http-requests/latency/patientAll/all"
"http-requests/count/coverageAll/dpc"
"http-requests/count/metadata/bcda"
"http-requests/count/all/ab2d"
"http-requests/count-not-2xx/all"
"http-requests/count/metadata/dpc"
"http-requests/count/metadata/ab2d"
"http-requests/count/patientAll/ab2d"
"http-requests/count/coverageAll/bb"
"http-requests/count-500/bcda"
"http-requests/count-not-2xx/bcda"
"http-requests/count/patientAll/dpc"
"http-requests/count-not-2xx/bb"
"http-requests/count/all/all"
"http-requests/count/coverageAll/bcda"
"http-requests/count/all/bcda"
"http-requests/count/eobAll/dpc"
"http-requests/count/patientAll/bb"
"http-requests/count/patientAll/all"
"http-requests/count/eobAll/bb"
"http-requests/latency/eobAll/all"
"http-requests/latency/metadata/all"
"http-requests/count-not-2xx/dpc"
"http-requests/count/all/dpc"
"http-requests/count/eobAll/all"
"http-requests/count/metadata/bb"
"http-requests/count/all/bb"
"http-requests/count/eobAll/bcda"
"http-requests/count/metadata/all"
"http-requests/count-500/ab2d"
"txt/http-requests/count-500/ab2d"
"txt/http-requests/latency/metadata/all"
"txt/http-requests/count/coverageAll/dpc"
"txt/http-requests/count/metadata/bb"
"txt/http-requests/count/coverageAll/bcda"
"txt/http-requests/count-500/bcda"
"txt/http-requests/count/patientAll/bcda"
"txt/http-requests/count/coverageAll/all"
"txt/http-requests/latency/coverageAll/all"
"txt/http-requests/count-not-2xx/bb"
"txt/http-requests/count/patientAll/all"
"txt/http-requests/count/patientAll/ab2d"
"txt/http-requests/count/all/all"
"txt/http-requests/count-500/bb"
"txt/http-requests/count/metadata/all"
"txt/http-requests/count/eobAll/bcda"
"txt/http-requests/count/patientAll/dpc"
"txt/http-requests/count/metadata/ab2d"
"txt/http-requests/count/metadata/dpc"
"txt/http-requests/latency/all/all"
"txt/http-requests/count-not-2xx/ab2d"
"txt/http-requests/count/patientAll/bb"
"txt/http-requests/count/eobAll/bb"
"txt/http-requests/count/all/bb"
"txt/http-requests/count-not-2xx/all"
"txt/http-requests/count/all/bcda"
"txt/http-requests/count/eobAll/dpc"
"txt/http-requests/count-not-2xx/dpc"
"txt/http-requests/count-not-2xx/bcda"
"txt/http-requests/count/eobAll/ab2d"
"txt/http-requests/count/coverageAll/bb"
"txt/http-requests/count/all/dpc"
"txt/http-requests/count/eobAll/all"
"txt/http-requests/count/metadata/bcda"
"txt/http-requests/count/all/ab2d"
"txt/http-requests/count-500/all"
"txt/http-requests/count/coverageAll/ab2d"
"txt/http-requests/latency/eobAll/all"
"txt/http-requests/count-500/dpc"
"txt/http-requests/latency/patientAll/all"
"http-requests/latency/coverageAll/all"
"http-requests/count-500/dpc"
```  
 - A new temp dashboard plotting both metric values and compared numbers was mocked out a metric points checked 
   -  Screen Shot 2022-07-28 at 3 08 00 PM
 - Verified new log stream reflect the filter pattern change in the metric filter name
   - Screen Shot 2022-07-28 at 3 09 01 PM
 - Verified old log stream reflect the `txt/` change in the metric filter name and metric name 
   -   Screen Shot 2022-07-28 at 3 09 14 PM
### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by mjburling — CHANGES_REQUESTED** [review]
This solution looks sound! 

1. I'm not sure how these survived this long, but it'd be great if those JSON-styled maps were formatted in HCL to keep things tidy. I've commented in-line on the first such map.
2. This is really wanting for a more thorough comparison between what already exists and the new resources. Beyond a successful plan/apply, I'm curious about the steps you've taken to evaluate these changes. As discussed, that's probably difficult to do in `test` without collecting (and potentially visualizing) the metrics associated with our internal traffic–could you start collecting those metrics? Filtering on e.g. `*data-server-client-test` should be appropriate for all environments. And to be clear, collecting metrics for _internal_ traffic should be an enduring change and won't need to be reversed.

File: ops/terraform/modules/resources/bfd_server_metrics/main.tf [hunk 1]
```diff
@@ -1,71 +1,91 @@
...
+    access = "/bfd/${var.env}/bfd-server/access.json"
   }
-
+  namespace = "bfd-${var.env}/bfd-server"
   endpoints = {
+    "all" : "*/fhir/*",
+    "metadata" : "*/fhir/metadata*",
+    "coverageAll" : "*/fhir/Coverage*",
+    "patientAll" : "*/fhir/Patient*",
+    "eobAll" : "*/fhir/ExplanationOfBenefit*",
```
[the remark this answers is not in the document]
Comment (kcirtapfromspace) [reply ops/terraform/modules/resources/bfd_server_metrics/main.tf]: I updated the one that would stay in place long term & left the tech debt in place mostly as is with a todo delete.

[the remark this answers is not in the document]
Comment (kcirtapfromspace) [reply ops/terraform/modules/resources/bfd_server_metrics/main.tf]: As for the second items I updated the PR comment with the validations that were performed. As old values persist the existing metrics namespace they can be validated against at any point with simple changes to the dashboard.

**Review by mjburling — COMMENTED** [review]
I think this might work just fine!

Ahead of accepting/merging, I'm still concerned about the 60 metric filters already attached to the `/bfd/prod-sbx/bfd-server/access.json` log group–I haven't looked _too_ closely, but I think they'll give us some grief at deployment time. Could you go ahead and remove those filters to avoid collisions?

**Comment by kcirtapfromspace** [comment]
Cleaned up existing log filters on sandbox. 
```
❯ aws logs  describe-metric-filters --log-group-name /bfd/prod-sbx/bfd-server/access.json | jq '.'
{
  "metricFilters": []
}
```

**Review by mjburling — APPROVED** [review]

**Review by christrotteradhoc — APPROVED** [review]

**Pull request merged**